app/src/main/java/com/hackadroid/pingu/com/hackadroid/pingu/activity/Intro3.java

```diff
@@ -0,0 +1,226 @@
+
+
+        locationCriteria = new Criteria();
+        locationCriteria.setAccuracy(Criteria.ACCURACY_COARSE);
+        locationCriteria.setAltitudeRequired(false);
+        locationCriteria.setBearingRequired(false);
+        locationCriteria.setCostAllowed(true);
+        locationCriteria.setPowerRequirement(Criteria.NO_REQUIREMENT);
+
+
+    /*    if (!manager.isProviderEnabled(LocationManager.GPS_PROVIDER)) {
+            Toast.makeText(getApplicationContext(), "GPS is disabled!Please turn on ", Toast.LENGTH_LONG).show();
+            Intent myIntent = new Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS);
+            startActivity(myIntent);
+        } else {
+            Toast.makeText(getApplicationContext(), "GPS is enabled!", Toast.LENGTH_LONG).show();
+            if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
+                // TODO: Consider calling
+                //    ActivityCompat#requestPermissions
+                // here to request the missing permissions, and then overriding
+                //   public void onRequestPermissionsResult(int requestCode, String[] permissions,
+                //                                          int[] grantResults)
+                // to handle the case where the user grants the permission. See the documentation
+                // for ActivityCompat#requestPermissions for more details.
+                return;
+            }
+            manager.requestLocationUpdates(providerName, 20000, 100, locationListener);
+        }*/
+
+        try {
+
+            GPSoff = Settings.Secure.getInt(getContentResolver(), Settings.Secure.LOCATION_MODE);
+
+        } catch (Settings.SettingNotFoundException e) {
+            e.printStackTrace();
+        }
+        if (GPSoff == 0) {
+            showMessageOKCancel("You need to turn Location On",
+                    new DialogInterface.OnClickListener() {
+                        @Override
+                        public void onClick(DialogInterface dialog, int which) {
+                            Intent onGPS = new Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS);
+                            startActivity(onGPS);
+                        }
+                    });
+        } else {
+            Toast.makeText(getApplicationContext(), "GPS is enabled!", Toast.LENGTH_LONG).show();
+        }
+
+
+
+    }
+    private void showMessageOKCancel(String message, DialogInterface.OnClickListener okListener) {
+        new AlertDialog.Builder(Intro3.this)
+                .setMessage(message)
+                .setPositiveButton("OK", okListener)
+                .setNegativeButton("Cancel", null)
+                .create()
+                .show();
+    }
+
+
+    private void findloc() {
+ /* GPS */
+        String bestprovider;
```

Comment:
I see this function does so many things, why can't it be broken into multiple.

Few suggestions : 

(Taken from Internal Answers/ Internet)

    When is returning "void" useful -

    If the method is returning a value only to indicate an incorrect state/exception, then it should be returning "void". The Exception should take care of handling all the error scenarios rather than a say a "boolean" return value. Now, the calling code needs to again validate both the boolean and the exception. If it had thrown only Exception, then it is much easier & clearer for the calling code.

    When is returning void not so useful -

    If the method performs any operation and returns a value as a result, then "void" should not be returned, the relevant result type(boolean/integer/object etc.) needs to be returned.

In general, return value should actual correspond to whether the method returns something or not. It should not correspond to whether the method executed right (well, it is supposed to). All the input validation/exception scenarios should be handled by means of throwing corresponding exceptions. Example, throw IllegalArgumentException when input is invalid, rather than returning "boolean" saying method did not execute.